File: tutorials/tutorial.py
"""Current pycasso interface examples for every supported family.

Run after installing the development checkout described in
``python-package/README.rst``.
"""

import numpy as np

import pycasso


rng = np.random.default_rng(7)
n, d = 180, 30
X = rng.normal(size=(n, d))


# Gaussian: the public default resolves the residual/covariance backend.
y_gaussian = 0.4 + X[:, 0] - 0.6 * X[:, 1] + rng.normal(scale=0.7, size=n)
gaussian = pycasso.Solver(
    X, y_gaussian, family="gaussian", penalty="l1",
    lambdas=(20, 0.05), type_gaussian="auto")
gaussian.train()
print("Gaussian backend:", gaussian.type_gaussian)
print("Gaussian beta shape:", gaussian.coef()["beta"].shape)
print("Gaussian prediction:",
      gaussian.predict(X[:3], lambdidx=gaussian.nlambda - 1))


# Binomial MCP: adaptive LLA may continue beyond its default three stages.
logit = -0.2 + 0.8 * X[:, 0] - 0.5 * X[:, 1]
y_binomial = rng.binomial(1, 1.0 / (1.0 + np.exp(-logit)))
binomial = pycasso.Solver(
    X, y_binomial, family="binomial", penalty="mcp",
    lambdas=(16, 0.1), lla_max_stages=5, fast_mode=True)
binomial.train()
print("Binomial status:", binomial.coef()["status"])
print("Binomial probabilities:",
      binomial.predict(X[:3], lambdidx=binomial.nlambda - 1))
print("Binomial confusion:",
      binomial.confusion(
          X, y_binomial, lambdidx=binomial.nlambda - 1)[0])


# Poisson: offsets are added on the link scale and required for new rows.
exposure = rng.uniform(0.5, 2.0, size=n)
offset = np.log(exposure)
poisson_mean = np.exp(offset + 0.1 + 0.25 * X[:, 0])
y_poisson = rng.poisson(poisson_mean)
poisson = pycasso.Solver(
    X, y_poisson, family="poisson", offset=offset,
    lambdas=(16, 0.1), fast_mode=True)
poisson.train()
print("Poisson means:",
      poisson.predict(
          X[:3], lambdidx=poisson.nlambda - 1,
          newoffset=offset[:3]))


# Square-root-lasso.
y_sqrt = X[:, 0] - X[:, 2] + rng.normal(size=n)
sqrt_model = pycasso.Solver(
    X, y_sqrt, family="sqrtlasso", lambdas=(16, 0.1))
sqrt_model.train()
sqrt_metrics = sqrt_model.assess()
print("Square-root-lasso final MSE:", sqrt_metrics["mse"][-1])


# Multinomial: original string labels are retained in class prediction.
scores = np.column_stack((
    0.8 * X[:, 0],
    -0.5 * X[:, 0] + 0.7 * X[:, 1],
    -0.6 * X[:, 1],
))
labels = np.array(["red", "green", "blue"])
y_multinomial = labels[np.argmax(
    scores + rng.normal(scale=0.5, size=scores.shape), axis=1)]
if np.unique(y_multinomial).size != 3:
    y_multinomial[:3] = labels

multinomial = pycasso.Solver(
    X, y_multinomial, family="multinomial", penalty="l1",
    lambdas=(16, 0.1), fast_mode=True)
multinomial.train()
print("Multinomial beta shape:", multinomial.coef()["beta"].shape)
print("Multinomial classes:",
      multinomial.predict(
          X[:5], lambdidx=multinomial.nlambda - 1, type="class"))


# Assessment and CV evaluate the retained path. Python indices are zero-based.
print("Multinomial class error:",
      multinomial.assess()["class_error"][-1])
cv = gaussian.cross_validate(nfolds=3, type_measure="deviance")
print("CV lambda.min:", cv["lambda_min"])
print("CV lambda.1se:", cv["lambda_1se"])

# Plotting is optional and requires Matplotlib:
# gaussian.plot(max_features=8)

File: tutorials/tutorial.R
library(picasso)

set.seed(20260719)
n <- 120L
d <- 24L
X <- matrix(rnorm(n * d), nrow = n)
beta <- c(1.2, -0.8, 0.5, rep(0, d - 3L))

## Gaussian: the public default chooses a memory-bounded backend automatically.
y.gaussian <- drop(0.4 + X %*% beta + rnorm(n, sd = 0.7))
fit.gaussian <- picasso(
  X, y.gaussian, family = "gaussian", method = "l1", nlambda = 16L
)
cat(
  "Gaussian backend:", fit.gaussian$type.gaussian.requested, "->",
  fit.gaussian$type.gaussian, "\n"
)
coef(
  fit.gaussian,
  lambda.idx = c(1L, fit.gaussian$nlambda),
  beta.idx = 1:4
)
predict(
  fit.gaussian, X[1:6, , drop = FALSE],
  s = tail(fit.gaussian$lambda, 1L),
  Y.pred.idx = 1:6
)
assess.picasso(fit.gaussian, X, y.gaussian)

## Explicit Gaussian backends remain available for reproducible comparisons.
fit.gaussian.naive <- picasso(
  X, y.gaussian, nlambda = 16L, type.gaussian = "naive"
)
fit.gaussian.covariance <- picasso(
  X, y.gaussian, nlambda = 16L, type.gaussian = "covariance"
)

## Binomial: MCP/SCAD use adaptive LLA around active-set Newton subproblems.
eta.binomial <- drop(X %*% beta)
y.binomial <- factor(
  ifelse(runif(n) < plogis(eta.binomial), "case", "control")
)
fit.binomial <- picasso(
  X, y.binomial, family = "binomial", method = "mcp",
  nlambda = 16L, lla.max.stages = 3L, fast.mode = TRUE
)
fit.binomial$status
head(fit.binomial$diagnostics)
predict(
  fit.binomial, X[1:6, , drop = FALSE],
  lambda.idx = fit.binomial$nlambda, type = "response",
  p.pred.idx = 1:6
)

## Poisson offsets are added on the link scale and must be supplied again for
## new-data prediction or assessment.
exposure <- runif(n, 0.5, 2.0)
offset <- log(exposure)
y.poisson <- rpois(n, exp(offset + 0.2 + X[, 1L] - 0.5 * X[, 2L]))
fit.poisson <- picasso(
  X, y.poisson, family = "poisson", nlambda = 16L, offset = offset
)
predict(
  fit.poisson, X[1:6, , drop = FALSE],
  lambda.idx = fit.poisson$nlambda, type = "response",
  newoffset = offset[1:6], p.pred.idx = 1:6
)
assess.picasso(fit.poisson, X, y.poisson, newoffset = offset)

## Square-root lasso uses active-set quadratic-MM updates and shares the
## adaptive-LLA diagnostics interface.
fit.sqrt <- picasso(
  X, y.gaussian, family = "sqrtlasso", nlambda = 16L
)
fit.sqrt$status

## Multinomial: labels are retained and probabilities are class-coupled.
scores <- cbind(
  red = X[, 1L] + 0.3 * X[, 2L],
  green = -X[, 1L] + 0.4 * X[, 3L],
  blue = -0.5 * X[, 2L] - 0.4 * X[, 3L]
)
y.multinomial <- factor(colnames(scores)[max.col(
  scores + matrix(rnorm(length(scores), sd = 0.5), nrow = n)
)])
fit.multinomial <- picasso(
  X, y.multinomial, family = "multinomial", nlambda = 16L,
  fast.mode = TRUE
)
predict(
  fit.multinomial, X[1:6, , drop = FALSE],
  lambda.idx = fit.multinomial$nlambda, type = "response"
)
predict(
  fit.multinomial, X[1:6, , drop = FALSE],
  lambda.idx = fit.multinomial$nlambda, type = "class"
)
confusion.picasso(
  fit.multinomial, X, y.multinomial,
  lambda.idx = fit.multinomial$nlambda
)

## Cross-validation uses one common lambda path. Categorical folds are
## stratified automatically; R fold IDs are one-based when supplied manually.
cv.gaussian <- cv.picasso(
  X, y.gaussian, family = "gaussian", nlambda = 16L, nfolds = 3L
)
cv.gaussian$lambda.min
coef(cv.gaussian, s = "lambda.1se", beta.idx = 1:4)

## Coefficient-path plots are available for every family.
plot(fit.gaussian)
plot(fit.multinomial, which.class = 1L)
